x/authz/keeper/keeper.go

```diff
@@ -0,0 +1,191 @@
+				err = k.update(ctx, grantee, granter, updated)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		handler := k.router.Handler(serviceMsg.Route())
+
+		if handler == nil {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized message route: %s", serviceMsg.Route())
+		}
+
+		msgResult, err = handler(ctx, serviceMsg.Request)
+		if err != nil {
+			return nil, sdkerrors.Wrapf(err, "failed to execute message; message %s", serviceMsg.MethodName)
+		}
+	}
+
+	return msgResult, nil
+}
+
+// Grant method grants the provided authorization to the grantee on the granter's account with the provided expiration
+// time. If there is an existing authorization grant for the same `sdk.Msg` type, this grant
+// overwrites that.
+func (k Keeper) Grant(ctx sdk.Context, grantee, granter sdk.AccAddress, authorization types.Authorization, expiration time.Time) error {
+	store := ctx.KVStore(k.storeKey)
+
+	grant, err := types.NewAuthorizationGrant(authorization, expiration)
+	if err != nil {
+		return err
+	}
+
+	bz := k.cdc.MustMarshalBinaryBare(&grant)
+	actor := types.GetActorAuthorizationKey(grantee, granter, authorization.MethodName())
+	store.Set(actor, bz)
+	return nil
+}
+
+// Revoke method revokes any authorization for the provided message type granted to the grantee by the granter.
+func (k Keeper) Revoke(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, msgType string) error {
+	store := ctx.KVStore(k.storeKey)
+	actor := types.GetActorAuthorizationKey(grantee, granter, msgType)
+	_, found := k.getAuthorizationGrant(ctx, actor)
+	if !found {
+		return sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "authorization not found")
+	}
+	store.Delete(actor)
+
+	return nil
+}
+
+// GetAuthorizations Returns list of `Authorizations` granted to the grantee by the granter.
+func (k Keeper) GetAuthorizations(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress) (authorizations []types.Authorization) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetActorAuthorizationKey(grantee, granter, "")
+	iter := sdk.KVStorePrefixIterator(store, key)
+	defer iter.Close()
+	var authorization types.AuthorizationGrant
+	for ; iter.Valid(); iter.Next() {
+		k.cdc.MustUnmarshalBinaryBare(iter.Value(), &authorization)
+		authorizations = append(authorizations, authorization.GetAuthorizationGrant())
+	}
+	return authorizations
+}
+
+// GetAuthorization Returns any `Authorization` (or `nil`), with the expiration time,
+// granted to the grantee by the granter for the provided msg type.
+func (k Keeper) GetAuthorization(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, msgType string) (cap types.Authorization, expiration time.Time) {
+	grant, found := k.getAuthorizationGrant(ctx, types.GetActorAuthorizationKey(grantee, granter, msgType))
+	if !found {
+		return nil, time.Time{}
+	}
+	if grant.Expiration.Before(ctx.BlockHeader().Time) {
+		k.Revoke(ctx, grantee, granter, msgType)
+		return nil, time.Time{}
+	}
```

Comment:
I don't think we should revoke in a method called `GetAuthorization`. I think having it in DispatchActions is enough.